### DISCUSSION TITLE
Add minor release tags for update automation

### DISCUSSION THREAD

**Discussion by emperorkebab** [zitadel/zitadel#8732] [Ideas]
Currently container releases with specific version tag or stable/release/debug
When automating container updates there can be breakages because minor updates introduce sometimes configuration changes that have to be done manually.

The idea is to have something like a latest tag but for each minor version, so that automation only updates patch releases
For example:
v2.63.3/v2.63/latest
v2.63.2
v2.63.1
v2.62.5/v2.62/stable
v2.63.0
v2.62.4
v2.62.3
v2.61.5/v2.61
v2.62.2
...

In container orchestration we will use the ghcr.io/zitadel/zitadel:v2.63 and automate patch updates without risks

**Comment by hifabienne** [comment]
@eliobischof @stebenz whats your thought on this?

**Reply by fforootd** [reply]
So what you mean is that when you deploy 2.63 for example it should always pull the latest patch version of that minor?
That would require imagepull always though and could impact your availabilty 😁 

We had such ideas in the past but abstained from them because it kind of violates the idea that an image tag should be immutable.

Happy to discuss this here!

**Reply by emperorkebab** [reply]
Hi, yes updated with image pull and availability isn't a concern with rolling updates

Correct me if i'm wrong but isn't this the same behavior for the latest and stable tags? I'm also using this tagging system as provided by a lot of other back-end service developers (like: [postgres](https://hub.docker.com/_/postgres), [huggingface chat-ui](https://github.com/huggingface/chat-ui/pkgs/container/chat-ui), [caddy](https://hub.docker.com/_/caddy), [home-assistant](https://github.com/home-assistant/core/pkgs/container/home-assistant/versions?filters%5Bversion_type%5D=tagged), [authelia](https://github.com/authelia/authelia/pkgs/container/authelia/versions?filters%5Bversion_type%5D=tagged), [devcontainers](https://github.com/devcontainers/features/pkgs/container/features%2Fnode/versions?filters%5Bversion_type%5D=tagged), [mailu containers](https://github.com/Mailu/Mailu/pkgs/container/clamav/versions?filters%5Bversion_type%5D=tagged), ... )
The reason this is needed and done by some devs out there is because docker registry doesn't have a way to pin to major/minor/patch version numbers like it's done for example in a node npm package.json

**Reply by fforootd** [reply]
> Hi, yes updated with image pull and availability isn't a concern with rolling updates

Well yeah in many cases it's not a big issue but we have seen problems when container or servers crash during rollout of a new version and we were unable to pull the new/old image 😂 but that is a corner case.

In regard of latest. That is a practice from docker and I do not like it that much. We only use this for examples and will not advise to use that for a production version.

Stable is an experiment to bring a stable version to customers but we already had some negative experiences with that, where customers made a copy of the tag and not updated it in their own registry.

There are definitely some projects who do the major/minor scheme but I am not so sure about it. So far our stance has been that customers should have a clear view on what they are running as a version.

> When automating container updates there can be breakages because minor updates introduce sometimes configuration changes that have to be done manually.

Can you elaborate this a little? With Zitadel you can upgrade minor/patch versions without breaking stuff (only bugs might affect this) 😝

Maybe at @eliobischof or @livio-a also have some thoughts to share on this.

**Reply by emperorkebab** [reply]
> Can you elaborate this a little?

In a scenario where the container is set with a stable image tag, automated updates will at times pull an image with a minor version bump, this is unwanted because minor (and major) versions introduce sometimes changes to the config/setup that have to be applied manually.
Updating versions manually using an exact version tag isn't exactly a solution as it represents a great amount of repetitive work, especially when having multiple services and databases, very frequently.

My postgres containers can be set with the tag 16, automation will take care of minor updates, fixes, and security patches without problems, my intervention is minimal as manual reviews of new major versions & updates are orders of magnitude less frequent, allowing me to keep up with patches without risking auto-upgrading to a version that needs a careful migration process.

I would like to add that for an identity management system and similar critical back-end services, security patches and fixes are obviously not negligible, without update automation it's easy to sometimes update the back-end with a delay that can be enough for preying malicious actors (who by the way, also use automation to try to quickly find un-patched vulnerable systems)

**Reply by livio-a** [reply]
> > Can you elaborate this a little?
> 
> In a scenario where the container is set with a stable image tag, automated updates will at times pull an image with a minor version bump, this is unwanted because minor (and major) versions introduce sometimes changes to the config/setup that have to be applied manually. Updating versions manually using an exact version tag isn't exactly a solution as it represents a great amount of repetitive work, especially when having multiple services and databases, very frequently.
> 

I think what @emperorkebab is referring to is the technical advisories (https://zitadel.com/docs/support/technical_advisory), where configuration changes might be needed, depending on the setup.

I've checked what others do and some actually publish a latest version of major / minor releases, e.g. keycloak, cockroach, while gitlab doesn't.

I mean it shouldn't take too much to do this, but the question is more on the guarantees that are expected from this. As @emperorkebab describes that no configuration changes are expected between patch versions. That might still happen at some point and more importantly there might be breaking behavior changes (for the end user) because of a bug fix. I mean every bug fix is a breaking behavior change, but we've run into situations where clients used and API in a manner, that was not documented, never intended and we were not aware of and after a supposedly simple fix, their system broke.

I personally would never set the image policy to always pull, even with rolling updates as any event leading the current container to stop might introduce an unintended update. I'd prefer more control of the version / state and upgrading on purpose.

Not having the automation on the pull policy would mean that at best I have some release channels or versioning, where as an operator of the system, I'm informed about new version (as with gitlab): https://github.com/zitadel/zitadel/issues/151

**Reply by emperorkebab** [reply]
> I mean every bug fix is a breaking behavior change, but we've run into situations where clients used and API in a manner, that was not documented, never intended and we were not aware of and after a supposedly simple fix, their system broke.

I think that taking consideration of breaking changes in version updates, is only applicable to intended behavior of zitadel, usage of unintended behavior is understandably uncharted territory that falls on the responsibility of us users.
But i totally understand if it's your concern, in this case the fix could be included in the next minor version (as is customary in other software release policies)

My proposition assumed that zitadel used [semver2](https://designsystem.morningstar.com/getting-started/versioning-and-breaking-changes/), to take its full advantage in the regards of the patching reactivity of automation.

> I think what @emperorkebab is referring to is the technical advisories (https://zitadel.com/docs/support/technical_advisory), where configuration changes might be needed, depending on the setup.

Not just that, but anything that introduces breaking change of intended behaviour that involves manual intervention, to move it out of automation's reach (this is in line with semver2's definition of patch release "[backward compatible bug fixes](https://semver.org/)")

**Reply by fforootd** [reply]
ATM we use semantic [versioning 2 as defined here ](https://semver.org/spec/v2.0.0.html).

In regard of release strategy. We include fixes in the next major/minor relases already but sometimes also backport patches if critical enough.

**Reply by emperorkebab** [reply]
That's good then, So the expectation from minor version tags doesn't conflict with the release strategy, i see no problem personally

**Reply by fforootd** [reply]
So the only question then left is to discuss if we want to have tags like 2, 2.63, 2.63.4 (just as an example) 😁

**Reply by emperorkebab** [reply]
As a follow up:
- [This Github action](https://github.com/marketplace/actions/docker-semver-publish) can auto-publish images with correct semver using "[`semver_higher`](https://github.com/marketplace/actions/docker-semver-publish#semver_higher)", it might help.
- This [medium article](https://medium.com/@mccode/using-semantic-versioning-for-docker-image-tags-dfde8be06699) probably better explaining semver tags than i did.

**Reply by fforootd** [reply]
Thanks for the added context.

Yeah I understand the general process. @eliobischof I think we should talk about this once you are back from vacation since we started a discussion on versioning a while ago 😁

**Reply by eliobischof** [reply]
I created the issue #9302  to follow-up on this
